lang/src/loader_account.rs

```diff
@@ -29,42 +28,50 @@ use std::ops::DerefMut;
 pub struct AccountLoader<'info, T: ZeroCopy + Owner> {
     acc_info: AccountInfo<'info>,
     phantom: PhantomData<&'info T>,
+    data: *mut T,
 }
 
 impl<'info, T: ZeroCopy + Owner + fmt::Debug> fmt::Debug for AccountLoader<'info, T> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("AccountLoader")
             .field("acc_info", &self.acc_info)
             .field("phantom", &self.phantom)
+            .field("data", unsafe { &*self.data })
             .finish()
     }
 }
 
 impl<'info, T: ZeroCopy + Owner> AccountLoader<'info, T> {
-    fn new(acc_info: AccountInfo<'info>) -> AccountLoader<'info, T> {
-        Self {
+    fn new(acc_info: AccountInfo<'info>) -> Result<AccountLoader<'info, T>, ProgramError> {
+        let data = bytemuck::from_bytes::<T>(&mut (acc_info.try_borrow_mut_data())?[8..])
+            as *const T as *mut T;
+
+        Ok(Self {
             acc_info,
             phantom: PhantomData,
-        }
+            data,
+        })
     }
 
     /// Constructs a new `Loader` from a previously initialized account.
     #[inline(never)]
     pub fn try_from(
         acc_info: &AccountInfo<'info>,
     ) -> Result<AccountLoader<'info, T>, ProgramError> {
-        if acc_info.owner != &T::owner() {
-            return Err(ErrorCode::AccountNotProgramOwned.into());
-        }
-        let data: &[u8] = &acc_info.try_borrow_data()?;
-        // Discriminator must match.
-        let mut disc_bytes = [0u8; 8];
-        disc_bytes.copy_from_slice(&data[..8]);
-        if disc_bytes != T::discriminator() {
-            return Err(ErrorCode::AccountDiscriminatorMismatch.into());
+        {
+            if acc_info.owner != &T::owner() {
+                return Err(ErrorCode::AccountNotProgramOwned.into());
+            }
+            let data: &[u8] = &acc_info.try_borrow_data()?;
+            // Discriminator must match.
+            let mut disc_bytes = [0u8; 8];
+            disc_bytes.copy_from_slice(&data[..8]);
+            if disc_bytes != T::discriminator() {
+                return Err(ErrorCode::AccountDiscriminatorMismatch.into());
+            }
```

Comment:
Why extra brackets?

Reply:
to get around borrowing issues with the borrowed account data. try_from uses new which uses borrow_mut which will fail without these braces because of the borrow in try_from still being active

Reply:
ah, I see. We also can use `drop(data)` which is probably more clear.
Thanks :+1:

Reply:
yea I tried that first but it didnt work so I went with this. would prefer drop and should just add a comment to it too. I probably messed up. will try with drop again

Reply:
ok so I adjusted it and now drop works but interestingly I had to make it 
`let data = acc_info.try_borrow_data()?;` (data is now of type Ref<&mut [u8]>) which to me was quite unexpected.

any idea why dropping data when using this line`let data: &[u8] = &acc_info.try_borrow_data()?;` does not work? I guess it's because data points to a `Ref` that is immediately referenced and therefore stays in limbo even when data is dropped?

Reply:
Yes, dropping `&[u8]` do not drop `Ref` so the previous code does not work.